# Sale Price prediction for bulldozers

### In this notebook I am  going to predict a sale price of bulldozers in the function of time and using given features.

### Problem definition:
*How good I'll be able to train a model that predict sale price of bulldozers using data and features downloaded from Kaggle.*

### Data:
Downloaded from:

Train and valid sets:

*https://www.kaggle.com/competitions/bluebook-for-bulldozers/data?select=TrainAndValid.csv*


*Train.csv is the training set, which contains data through the end of 2011.*

*Valid.csv is the validation set, which contains data from January 1, 2012 - April 30, 2012 You make predictions on this set throughout the majority of the competition. Your score on this set is used to create the public leaderboard.*



### Features: 
Downloaded from:
*https://www.kaggle.com/competitions/bluebook-for-bulldozers/data?select=Data+Dictionary.xlsx*

### Evaluation:
The evaluation metric for this competition is the RMSLE (root mean squared log error) between the actual and predicted auction prices.

#### *note:  my goal is to minimize RMSLE.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn
import datetime as dt
%matplotlib inline


In [125]:
# Loading the train_valid file.
train_valid = pd.read_csv('f:/TrainAndValid.csv',low_memory=False)

In [126]:
# Gerneral view of data structure.
train_valid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 412698 entries, 0 to 412697
Data columns (total 53 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   SalesID                   412698 non-null  int64  
 1   SalePrice                 412698 non-null  float64
 2   MachineID                 412698 non-null  int64  
 3   ModelID                   412698 non-null  int64  
 4   datasource                412698 non-null  int64  
 5   auctioneerID              392562 non-null  float64
 6   YearMade                  412698 non-null  int64  
 7   MachineHoursCurrentMeter  147504 non-null  float64
 8   UsageBand                 73670 non-null   object 
 9   saledate                  412698 non-null  object 
 10  fiModelDesc               412698 non-null  object 
 11  fiBaseModel               412698 non-null  object 
 12  fiSecondaryDesc           271971 non-null  object 
 13  fiModelSeries             58667 non-null   o

In [127]:
# General view on missing labels quantity.
train_valid.isna().sum()

SalesID                          0
SalePrice                        0
MachineID                        0
ModelID                          0
datasource                       0
auctioneerID                 20136
YearMade                         0
MachineHoursCurrentMeter    265194
UsageBand                   339028
saledate                         0
fiModelDesc                      0
fiBaseModel                      0
fiSecondaryDesc             140727
fiModelSeries               354031
fiModelDescriptor           337882
ProductSize                 216605
fiProductClassDesc               0
state                            0
ProductGroup                     0
ProductGroupDesc                 0
Drive_System                305611
Enclosure                      334
Forks                       214983
Pad_Type                    331602
Ride_Control                259970
Stick                       331602
Transmission                224691
Turbocharged                331602
Blade_Extension     

### Because we deal with time series data it is recommended  to parse the dates, so we have to import the data once again with parse dates parameter for chosen column.

In [128]:
# Import train_valid set with parse_dates parameter.
train_valid = pd.read_csv('f:/TrainAndValid.csv',low_memory=False,parse_dates=['saledate'])

###      We are going to copy of the original datasets so all of the changes would not affect oryginal datasets.

In [129]:
train_valid_copy = train_valid.copy()

### Sorting data by 'saledate' to have a better view af sales in a function of time.

In [130]:
train_valid_copy.sort_values(by=['saledate'],inplace=True)

In [131]:
train_valid_copy.saledate

205615   1989-01-17
274835   1989-01-31
141296   1989-01-31
212552   1989-01-31
62755    1989-01-31
            ...    
410879   2012-04-28
412476   2012-04-28
411927   2012-04-28
407124   2012-04-28
409203   2012-04-28
Name: saledate, Length: 412698, dtype: datetime64[ns]

### Check of actual data structure and types.

In [132]:
train_valid_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 412698 entries, 205615 to 409203
Data columns (total 53 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   SalesID                   412698 non-null  int64         
 1   SalePrice                 412698 non-null  float64       
 2   MachineID                 412698 non-null  int64         
 3   ModelID                   412698 non-null  int64         
 4   datasource                412698 non-null  int64         
 5   auctioneerID              392562 non-null  float64       
 6   YearMade                  412698 non-null  int64         
 7   MachineHoursCurrentMeter  147504 non-null  float64       
 8   UsageBand                 73670 non-null   object        
 9   saledate                  412698 non-null  datetime64[ns]
 10  fiModelDesc               412698 non-null  object        
 11  fiBaseModel               412698 non-null  object        
 1

In [133]:
train_valid_copy.isna().sum()

SalesID                          0
SalePrice                        0
MachineID                        0
ModelID                          0
datasource                       0
auctioneerID                 20136
YearMade                         0
MachineHoursCurrentMeter    265194
UsageBand                   339028
saledate                         0
fiModelDesc                      0
fiBaseModel                      0
fiSecondaryDesc             140727
fiModelSeries               354031
fiModelDescriptor           337882
ProductSize                 216605
fiProductClassDesc               0
state                            0
ProductGroup                     0
ProductGroupDesc                 0
Drive_System                305611
Enclosure                      334
Forks                       214983
Pad_Type                    331602
Ride_Control                259970
Stick                       331602
Transmission                224691
Turbocharged                331602
Blade_Extension     

### As we see we have mostly 'Object' type of data and a few numeric type.We can also see there is a big quantity of missing data.
### To start modelling we must change 'Object' type data into numbers ,as well as get sorted missing values. 

### One of ways to change 'object' / 'string' type data into numbers is to make them categorical type.


In [134]:
# Changing 'Object' type columns into category type.
for label,col in train_valid_copy.items():
    if pd.api.types.is_object_dtype(col):
        train_valid_copy[label] = col.astype('category')
        
    

In [135]:
train_valid_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 412698 entries, 205615 to 409203
Data columns (total 53 columns):
 #   Column                    Non-Null Count   Dtype         
---  ------                    --------------   -----         
 0   SalesID                   412698 non-null  int64         
 1   SalePrice                 412698 non-null  float64       
 2   MachineID                 412698 non-null  int64         
 3   ModelID                   412698 non-null  int64         
 4   datasource                412698 non-null  int64         
 5   auctioneerID              392562 non-null  float64       
 6   YearMade                  412698 non-null  int64         
 7   MachineHoursCurrentMeter  147504 non-null  float64       
 8   UsageBand                 73670 non-null   category      
 9   saledate                  412698 non-null  datetime64[ns]
 10  fiModelDesc               412698 non-null  category      
 11  fiBaseModel               412698 non-null  category      
 1

In [136]:
# Sorting data by "saledate" columns.
train_valid_copy.sort_values(by='saledate',inplace=True)

In [137]:
# Filling missing data for numeric type with median of numeric columns.
for label,col in train_valid_copy.items():
    if pd.api.types.is_numeric_dtype(col):
        if pd.isnull(col).sum():
            train_valid_copy[label] = col.fillna(col.median())

In [138]:
# Changing categorical type to numeric values:
for label,col in train_valid_copy.items():
    if label == 'saledate':
        continue
    if not pd.api.types.is_numeric_dtype(col):
        train_valid_copy[label] = pd.Categorical(col).codes + 1

In [139]:
train_valid_copy.isnull().sum()

SalesID                     0
SalePrice                   0
MachineID                   0
ModelID                     0
datasource                  0
auctioneerID                0
YearMade                    0
MachineHoursCurrentMeter    0
UsageBand                   0
saledate                    0
fiModelDesc                 0
fiBaseModel                 0
fiSecondaryDesc             0
fiModelSeries               0
fiModelDescriptor           0
ProductSize                 0
fiProductClassDesc          0
state                       0
ProductGroup                0
ProductGroupDesc            0
Drive_System                0
Enclosure                   0
Forks                       0
Pad_Type                    0
Ride_Control                0
Stick                       0
Transmission                0
Turbocharged                0
Blade_Extension             0
Blade_Width                 0
Enclosure_Type              0
Engine_Horsepower           0
Hydraulics                  0
Pushblock 

In [140]:
train_valid_copy.dtypes

SalesID                              int64
SalePrice                          float64
MachineID                            int64
ModelID                              int64
datasource                           int64
auctioneerID                       float64
YearMade                             int64
MachineHoursCurrentMeter           float64
UsageBand                             int8
saledate                    datetime64[ns]
fiModelDesc                          int16
fiBaseModel                          int16
fiSecondaryDesc                      int16
fiModelSeries                         int8
fiModelDescriptor                    int16
ProductSize                           int8
fiProductClassDesc                    int8
state                                 int8
ProductGroup                          int8
ProductGroupDesc                      int8
Drive_System                          int8
Enclosure                             int8
Forks                                 int8
Pad_Type   

## Now when we have only numerical values in our dataset,we are able to split set into train and valid sets separately as documentation says.

#### Train.csv is the training set, which contains data through the end of 2011.
#### Valid.csv is the validation set, which contains data from January 1, 2012 - April 30, 2012.

You make predictions on this set throughout the majority of the competition. Your score on this set is used to create the public leaderboard.

##### As it's written the split is dependent to the date,telling that untill the end of 2011 data should be in Train.csv 
##### and Valid.csv should be from 1.1.2012 - 30.04.2012.
##### By using our 'saledate' column we are able to split our train_valid.copy dataframe into train and test sets.

In [141]:
train_valid_copy.T

205615               233186  \
SalesID                               1646770              1728883   
SalePrice                              9500.0              30000.0   
MachineID                             1126363              1523610   
ModelID                                  8434                 9105   
datasource                                132                  132   
auctioneerID                             18.0                 99.0   
YearMade                                 1974                 1986   
MachineHoursCurrentMeter                  0.0                  0.0   
UsageBand                                   0                    0   
saledate                  1989-01-17 00:00:00  1989-01-31 00:00:00   
fiModelDesc                              4593                 4772   
fiBaseModel                              1744                 1852   
fiSecondaryDesc                             0                    0   
fiModelSeries                               0                    0   
fiModelDescriptor                           0                    0   
ProductSize                                 4                    0   
fiProductClassDesc                         47                   73   
state                                      44                    9   
ProductGroup                                5                    6   
ProductGroupDesc                            5                    6   
Drive_System                                0                    0   
Enclosure                                   6                    1   
Forks                                       0                    0   
Pad_Type                                    0                    0   
Ride_Control                                0                    0   
Stick                                       0                    0   
Transmission                                3                    0   
Turbocharged                                0                    0   
Blade_Extension                             0                    0   
Blade_Width                                 0                    0   
Enclosure_Type                              0                    0   
Engine_Horsepower                           0                    0   
Hydraulics                                  1                    0   
Pushblock                                   0                    0   
Ripper                                      2                    0   
Scarifier                                   0                    0   
Tip_Control                                 0                    0   
Tire_Size                                   0                   14   
Coupler                                     0                    0   
Coupler_System                              0                    0   
Grouser_Tracks                              0                    0   
Hydraulics_Flow                             0                    0   
Track_Type                                  0                    0   
Undercarriage_Pad_Width                     0                    0   
Stick_Length                                0                    0   
Thumb                                       0                    0   
Pattern_Changer                             0                    0   
Grouser_Type                                0                    0   
Backhoe_Mounting                            1                    0   
Blade_Type                                  8                    0   
Travel_Controls                             6                    0   
Differential_Type                           0                    0   
Steering_Controls                           0                    0   

                                       142491               115536  \
SalesID                               1508502              1452578   
SalePrice                             21000.0              33000.0   
MachineID                             1153157   

In [142]:
train_valid_copy['YearOfSale'] = train_valid_copy['saledate'].dt.year
train_valid_copy.drop(['saledate'],axis=1,inplace=True)

In [143]:
train_valid_copy.head().T

205615     233186     142491     115536  \
SalesID                   1646770.0  1728883.0  1508502.0  1452578.0   
SalePrice                    9500.0    30000.0    21000.0    33000.0   
MachineID                 1126363.0  1523610.0  1153157.0  1544443.0   
ModelID                      8434.0     9105.0     4138.0     3854.0   
datasource                    132.0      132.0      132.0      132.0   
auctioneerID                   18.0       99.0       99.0       99.0   
YearMade                     1974.0     1986.0     1972.0     1974.0   
MachineHoursCurrentMeter        0.0        0.0        0.0        0.0   
UsageBand                       0.0        0.0        0.0        0.0   
fiModelDesc                  4593.0     4772.0     2347.0     1767.0   
fiBaseModel                  1744.0     1852.0      713.0      535.0   
fiSecondaryDesc                 0.0        0.0       57.0       31.0   
fiModelSeries                   0.0        0.0        0.0        0.0   
fiModelDescriptor               0.0        0.0        0.0        0.0   
ProductSize                     4.0        0.0        2.0        4.0   
fiProductClassDesc             47.0       73.0       50.0       62.0   
state                          44.0        9.0        9.0        9.0   
ProductGroup                    5.0        6.0        5.0        6.0   
ProductGroupDesc                5.0        6.0        5.0        6.0   
Drive_System                    0.0        0.0        0.0        0.0   
Enclosure                       6.0        1.0        6.0        1.0   
Forks                           0.0        0.0        0.0        1.0   
Pad_Type                        0.0        0.0        0.0        0.0   
Ride_Control                    0.0        0.0        0.0        2.0   
Stick                           0.0        0.0        0.0        0.0   
Transmission                    3.0        0.0        8.0        0.0   
Turbocharged                    0.0        0.0        0.0        0.0   
Blade_Extension                 0.0        0.0        0.0        0.0   
Blade_Width                     0.0        0.0        0.0        0.0   
Enclosure_Type                  0.0        0.0        0.0        0.0   
Engine_Horsepower               0.0        0.0        0.0        0.0   
Hydraulics                      1.0        0.0        1.0        1.0   
Pushblock                       0.0        0.0        0.0        0.0   
Ripper                          2.0        0.0        2.0        0.0   
Scarifier                       0.0        0.0        0.0        0.0   
Tip_Control                     0.0        0.0        0.0        0.0   
Tire_Size                       0.0       14.0        0.0       12.0   
Coupler                         0.0        0.0        0.0        3.0   
Coupler_System                  0.0        0.0        0.0        0.0   
Grouser_Tracks                  0.0        0.0        0.0        0.0   
Hydraulics_Flow                 0.0        0.0        0.0        0.0   
Track_Type                      0.0        0.0        0.0        0.0   
Undercarriage_Pad_Width         0.0        0.0        0.0        0.0   
Stick_Length                    0.0        0.0        0.0        0.0   
Thumb                           0.0        0.0        0.0        0.0   
Pattern_Changer                 0.0        0.0        0.0        0.0   
Grouser_Type                    0.0        0.0        0.0        0.0   
Backhoe_Mounting                1.0        0.0        1.0        0.0   
Blade_Type                      8.0        0.0        1.0        0.0   
Travel_Controls                 6.0        0.0        6.0        0.0   
Differential_Type               0.0        0.0        0.0        4.0   
Steering_Controls               0.0        0.0        0.0        2.0   
YearOfSale                   1989.0     1989.0     1989.0     1989.0   

                             92301   
SalesID                   1403418.0  
SalePrice                   24000.0  
MachineID                 

In [144]:
train_valid_copy.tail().T

409901     405777     411889     411890  \
SalesID                   6286556.0  6256957.0  6312472.0  6312473.0   
SalePrice                   23000.0    11500.0    11000.0    11000.0   
MachineID                 1793587.0  1860454.0  1820676.0  1796787.0   
ModelID                      4605.0     6868.0    17231.0    17231.0   
datasource                    149.0      149.0      149.0      149.0   
auctioneerID                   99.0       99.0       99.0       99.0   
YearMade                     2006.0     1000.0     2006.0     2006.0   
MachineHoursCurrentMeter        0.0        0.0        0.0        0.0   
UsageBand                       0.0        0.0        0.0        0.0   
fiModelDesc                   500.0     1679.0      581.0      581.0   
fiBaseModel                   176.0      505.0      191.0      191.0   
fiSecondaryDesc                59.0        0.0        0.0        0.0   
fiModelSeries                   0.0        0.0        0.0        0.0   
fiModelDescriptor               0.0        0.0        0.0        0.0   
ProductSize                     0.0        3.0        5.0        5.0   
fiProductClassDesc              2.0       19.0        6.0        6.0   
state                           5.0        5.0        5.0        5.0   
ProductGroup                    1.0        4.0        4.0        4.0   
ProductGroupDesc                1.0        4.0        4.0        4.0   
Drive_System                    2.0        0.0        0.0        0.0   
Enclosure                       6.0        1.0        1.0        1.0   
Forks                           1.0        0.0        0.0        0.0   
Pad_Type                        2.0        0.0        0.0        0.0   
Ride_Control                    1.0        0.0        0.0        0.0   
Stick                           1.0        0.0        0.0        0.0   
Transmission                    8.0        0.0        0.0        0.0   
Turbocharged                    1.0        0.0        0.0        0.0   
Blade_Extension                 0.0        0.0        0.0        0.0   
Blade_Width                     0.0        0.0        0.0        0.0   
Enclosure_Type                  0.0        0.0        0.0        0.0   
Engine_Horsepower               0.0        0.0        0.0        0.0   
Hydraulics                      0.0       12.0        4.0        4.0   
Pushblock                       0.0        0.0        0.0        0.0   
Ripper                          0.0        0.0        0.0        0.0   
Scarifier                       0.0        0.0        0.0        0.0   
Tip_Control                     0.0        0.0        0.0        0.0   
Tire_Size                       0.0        0.0        0.0        0.0   
Coupler                         0.0        3.0        3.0        3.0   
Coupler_System                  0.0        0.0        0.0        0.0   
Grouser_Tracks                  0.0        0.0        0.0        0.0   
Hydraulics_Flow                 0.0        0.0        0.0        0.0   
Track_Type                      0.0        2.0        2.0        2.0   
Undercarriage_Pad_Width         0.0       19.0       19.0       19.0   
Stick_Length                    0.0       29.0       29.0       29.0   
Thumb                           0.0        2.0        3.0        3.0   
Pattern_Changer                 0.0        2.0        2.0        2.0   
Grouser_Type                    0.0        1.0        1.0        1.0   
Backhoe_Mounting                0.0        0.0        0.0        0.0   
Blade_Type                      0.0        0.0        0.0        0.0   
Travel_Controls                 0.0        0.0        0.0        0.0   
Differential_Type               0.0        0.0        0.0        0.0   
Steering_Controls               0.0        0.0        0.0        0.0   
YearOfSale                   2012.0     2012.0     2012.0     2012.0   

                             409203  
SalesID                   6283635.0  
SalePrice                   34000.0  
MachineID                 

### Now when all our values are numeric,we are able to fit them to a ML model.

In [145]:
TrainSet = train_valid_copy[train_valid_copy['YearOfSale'] != 2012]

In [146]:
ValidationSet = train_valid_copy[train_valid_copy['YearOfSale'] == 2012]

In [147]:
X_train = TrainSet.drop(['SalePrice'],axis=1)
y_train = TrainSet['SalePrice']
X_valid = ValidationSet.drop(['SalePrice'],axis=1)
y_valid = ValidationSet['SalePrice']

In [148]:
X_train

SalesID  MachineID  ModelID  datasource  auctioneerID  YearMade  \
205615  1646770    1126363     8434         132          18.0      1974   
233186  1728883    1523610     9105         132          99.0      1986   
142491  1508502    1153157     4138         132          99.0      1972   
115536  1452578    1544443     3854         132          99.0      1974   
92301   1403418    1390168     7110         132          99.0      1986   
...         ...        ...      ...         ...           ...       ...   
397551  6286845    1818840     8050         149           2.0      2006   
397589  6287076    1862704     4640         149           2.0      1000   
397660  6287313    1805275     4666         149           2.0      2004   
397995  6288403    1908296    26287         149           2.0      1000   
397022  6285369    1811309     4666         149           2.0      2004   

        MachineHoursCurrentMeter  UsageBand  fiModelDesc  fiBaseModel  ...  \
205615                       0.0          0         4593         1744  ...   
233186                       0.0          0         4772         1852  ...   
142491                       0.0          0         2347          713  ...   
115536                       0.0          0         1767          535  ...   
92301                        0.0          0          744          245  ...   
...                          ...        ...          ...          ...  ...   
397551                       0.0          0          493          176  ...   
397589                       0.0          0          732          242  ...   
397660                       0.0          0          865          267  ...   
397995                       0.0          0         1818          557  ...   
397022                       0.0          0          865          267  ...   

        Stick_Length  Thumb  Pattern_Changer  Grouser_Type  Backhoe_Mounting  \
205615             0      0                0             0                 1   
233186             0      0                0             0                 0   
142491             0      0                0             0                 1   
115536             0      0                0             0                 0   
92301              0      0                0             0                 0   
...              ...    ...              ...           ...               ...   
397551             0      0                0             0                 0   
397589             0      0                0             0                 0   
397660             0      0                0             0                 1   
397995             0      0                0             0                 0   
397022             0      0                0             0                 1   

        Blade_Type  Travel_Controls  Differential_Type  Steering_Controls  \
205615           8                6                  0                  0   
233186           0                0                  0                  0   
142491           1                6                  0                  0   
115536           0                0                  4                  2   
92301            0                0                  0                  0   
...            ...              ...                ...                ...   
397551           0                0                  0                  0   
397589           0                0                  0                  0   
397660           6                6                  0                  0   
397995           0                0                  0                  0   
397022           6                6                  0                  0   

        YearOfSale  
205615        1989  
233186        1989  
142491        1989  
115536        1989  
92301         1989  
...            ...  
397551        2011  
397589        2011  
397660        2011  
397995        2011  
397022        2011  

[401125 rows x 52 columns]

In [149]:
y_train

205615     9500.0
233186    30000.0
142491    21000.0
115536    33000.0
92301     24000.0
           ...   
397551    34500.0
397589    19000.0
397660    24000.0
397995    18000.0
397022    24000.0
Name: SalePrice, Length: 401125, dtype: float64

In [150]:
#Import Random Forest Regressor.
from sklearn.ensemble import RandomForestRegressor

In [183]:
#Instantiate and fit a model to a given data.


model = RandomForestRegressor(n_jobs=-1)
model.fit(X_train,y_train)


RandomForestRegressor(n_jobs=-1)

In [185]:
y_pred =model.predict(X_valid)

In [152]:
#model.score(X_valid,y_valid)

0.8739273829690865

In [188]:
#Our goal is to minimize the Root Mean Squared log Error...
#Import Mean Squared Log Error.
from sklearn.metrics import mean_squared_log_error

np.sqrt(mean_squared_log_error(y_valid,y_pred))

0.24921126566184412

##### The required score for Root Mean Squared Log Error:
0.24921126566184412